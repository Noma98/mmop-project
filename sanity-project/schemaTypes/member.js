export default {
  title: 'Member',
  name: 'member',
  type: 'document',
  fields: [
    {
      title: 'UserId',
      name: 'userId',
      type: 'string',
    },
    {
      title: 'UserName',
      name: 'userName',
      type: 'string',
    },
    {
      title: 'Email',
      name: 'email',
      type: 'string',
    },
    {
      title: 'PhoneNum',
      name: 'phoneNum',
      type: 'string',
    },
    {
      title: 'Profile',
      name: 'profile',
      type: 'image',
      validation: (Rule) => Rule.optional(),
    },
    {
      title: 'GoogleProfile',
      name: 'googleProfile',
      type: 'string',
    },
    {
      title: 'Projects',
      name: 'projects',
      type: 'array',
      of: [
        {
          type: 'reference',
          to: [{type: 'project'}],
        },
      ],
      validation: (Rule) => Rule.unique(),
    },
    {
      title: 'Setting',
      name: 'setting',
      type: 'reference',
      to: [{type: 'setting'}],
    },
    {
      title: 'Skills',
      name: 'skills',
      type: 'array',
      of: [
        {
          title: 'Skill',
          name: 'skill',
          type: 'string',
        },
      ],
    },
  ],
  preview: {
    select: {
      title: 'userId',
      subtitle: 'email',
    },
    prepare(selection) {
      return selection
    },
  },
}
