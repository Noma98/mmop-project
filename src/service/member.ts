import SanityClient from '@/service/sanity';
import SettingSanity from '@/service/setting';

export type OAuthMember = {
  id: string;
  userName: string;
  userId: string;
  email: string;
  image: string;
};
export type FullMember = {
  userId: string;
  userName?: string;
  phoneNum?: string;
  email: string;
  skills: string[];
  profile?: string;
  googleProfile: string;
  github?: string;
  setting?: {
    title: string;
    subtitle: string;
    logo?: string;
    bgColors: {
      left: string;
      right: string;
    };
    introduction: string;
  };
};
type FormMember = FullMember & { _id: string };

export interface MemberSanityType {
  create({ id, userId, email, userName, image }: OAuthMember): Promise<any>;
  read(userId: string): Promise<FullMember | null>;
  update(data: FormMember, file: File | null): Promise<any>;
}
export default class MemberSanity
  extends SanityClient
  implements MemberSanityType
{
  public create = async ({
    id,
    userId,
    email,
    userName,
    image,
  }: OAuthMember) => {
    const settingSanity = new SettingSanity();
    const newSetting = await settingSanity.create(id);
    const data = {
      _id: id,
      _type: 'member',
      userId,
      userName,
      email,
      phoneNum: '',
      googleProfile: image,
      projects: [],
      setting: {
        _ref: newSetting._id,
        _type: 'setting',
      },
      skills: [],
      github: '',
    };
    return this.getClient().createIfNotExists(data);
  };
  public read = async (userId: string): Promise<FullMember | null> => {
    const data = await this.getClient().fetch(
      `*[_type=="member"&&userId=="${userId}"][0]{...,setting->}`
    );
    return {
      ...data,
      profile: data.profile && this.urlFor(data.profile),
      setting: {
        ...data.setting,
        logo: !data.setting.logo ? '' : this.urlFor(data.setting.logo),
      },
    };
  };
  public update = async (data: FormMember, file: File | null) => {
    const { _id, userName, phoneNum, skills, profile, github } = data;
    return this.getClient()
      .patch(_id)
      .set({
        _type: 'member',
        profile:
          !profile && !file
            ? ''
            : {
                _type: 'image',
                asset: {
                  _type: 'reference',
                  _ref: file
                    ? (await this.getClient().assets.upload('image', file))._id
                    : this.extractAssetIdFromUrl(profile as string),
                },
              },
        userName,
        phoneNum,
        skills,
        github,
      })
      .commit();
  };
}
