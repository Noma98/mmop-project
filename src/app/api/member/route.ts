import { NextRequest, NextResponse } from 'next/server';

import { withSessionUser } from '@/utils/session';
import { sanityService } from '@/service';

export async function PUT(req: NextRequest) {
  return withSessionUser(async () => {
    const formData = await req.formData();
    const data = JSON.parse(formData.get('data')?.toString() as string);
    const file = (formData.get('file') as File) || null;

    return sanityService.member
      .update(data, file)
      .then((res) => NextResponse.json(res))
      .catch((err) => new Response(JSON.stringify(err), { status: 500 }));
  });
}
