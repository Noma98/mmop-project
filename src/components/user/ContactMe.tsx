import React from 'react';

import { FullMember } from '@/service/member';
import Wave from '@/components/common/Wave';
import { GithubIcon, GmailIcon, PhoneIcon } from '@/components/icons';

export default function ContactMe({
  email,
  phoneNum,
  userName,
  github,
}: FullMember) {
  const copyText = (text: string) => {
    navigator.clipboard
      .writeText(text)
      .then(() => alert('Copied it to your clipboard!'))
      .catch((err) => alert('Error occured.'));
  };

  return (
    <section className='w-full h-80 flex flex-col items-center py-32 pb-[500px] bg-point3'>
      <h2 className='highlight font-bold text-4xl mb-4 bg-point2 px-2 border-l-[6px] border-orange-300'>
        Contact me!
      </h2>
      <div className='my-4 border-[1px] p-10 w-[440px] flex flex-col items-start bg-white rounded-xl'>
        <p className='font-bold mb-4 text-xl '>{userName}</p>
        <p className='text-gray-700 text-md'>
          <b className='text-black text-lg mr-3'>E-mail </b>
          {email}
        </p>
        <p className='text-gray-700 text-md mt-1'>
          <b className='text-black text-lg mr-3'>Phone </b>
          {phoneNum?.replace(/(\d{3})(\d{4})(\d{4})/, '$1-$2-$3')}
        </p>
        <div className='flex gap-4 mt-8 ml-auto text-gray-800'>
          {github && (
            <a href={github} className='hover:text-point3'>
              <GithubIcon size='8' />
            </a>
          )}
          <button
            aria-label='email'
            onClick={() => copyText(email)}
            className='hover:text-point3'
          >
            <GmailIcon />
          </button>
          {phoneNum && (
            <button
              aria-label='phone'
              onClick={() => copyText(phoneNum)}
              className='hover:text-point3'
            >
              <PhoneIcon />
            </button>
          )}
        </div>
      </div>
      <Wave />
    </section>
  );
}
