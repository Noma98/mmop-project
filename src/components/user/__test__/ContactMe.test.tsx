import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ContactMe from '../ContactMe';
import { testMember } from '../../../tests/member';

describe('ContactMe', () => {
  beforeEach(() => {
    Object.assign(navigator, {
      clipboard: {
        writeText: jest.fn().mockResolvedValue(''),
      },
    });
    Object.assign(window, {
      alert: jest.fn(),
    });
  });
  it('renders correctly', () => {
    const { asFragment } = render(<ContactMe {...testMember} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders all buttons if all properties are provided', () => {
    render(<ContactMe {...testMember} />);
    expect(screen.queryByRole('link')).toHaveAttribute(
      'href',
      testMember.github
    );
    expect(screen.queryByLabelText('email')).toBeInTheDocument();
    expect(screen.queryByLabelText('phone')).toBeInTheDocument();
  });

  it('renders only email button if there is no github and phoneNum property', () => {
    render(
      <ContactMe {...testMember} github={undefined} phoneNum={undefined} />
    );
    expect(screen.queryByRole('link')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('email')).toBeInTheDocument();
    expect(screen.queryByLabelText('phone')).not.toBeInTheDocument();
  });

  it('copy email text when click email button', async () => {
    render(<ContactMe {...testMember} />);
    userEvent.click(screen.getByLabelText('email'));
    await waitFor(() =>
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
        testMember.email
      )
    );
  });
  it('copies phone number and show alert when clicking phone button', async () => {
    render(<ContactMe {...testMember} />);
    userEvent.click(screen.getByLabelText('phone'));
    await waitFor(() =>
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
        testMember.phoneNum
      )
    );
    expect(window.alert).toHaveBeenCalledWith('Copied it to your clipboard!');
  });
  it('show error alert when copying phone number fails', async () => {
    (navigator.clipboard.writeText as jest.Mock).mockRejectedValue(
      new Error('error')
    );
    render(<ContactMe {...testMember} />);
    userEvent.click(screen.getByLabelText('phone'));
    await waitFor(() =>
      expect(navigator.clipboard.writeText).toHaveBeenCalledWith(
        testMember.phoneNum
      )
    );
    expect(window.alert).toHaveBeenCalledWith('Error occured.');
  });
});
