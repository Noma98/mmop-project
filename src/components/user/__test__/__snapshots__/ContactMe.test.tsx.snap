// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactMe renders correctly 1`] = `
<DocumentFragment>
  <section
    class="w-full h-80 flex flex-col items-center py-32 pb-[500px] bg-point3"
  >
    <h2
      class="highlight font-bold text-4xl mb-4 bg-point2 px-2 border-l-[6px] border-orange-300"
    >
      Contact me!
    </h2>
    <div
      class="my-4 border-[1px] p-10 w-[440px] flex flex-col items-start bg-white rounded-xl"
    >
      <p
        class="font-bold mb-4 text-xl "
      >
        testName
      </p>
      <p
        class="text-gray-700 text-md"
      >
        <b
          class="text-black text-lg mr-3"
        >
          E-mail 
        </b>
        testId@gmail.com
      </p>
      <p
        class="text-gray-700 text-md mt-1"
      >
        <b
          class="text-black text-lg mr-3"
        >
          Phone 
        </b>
        010-1234-1234
      </p>
      <div
        class="flex gap-4 mt-8 ml-auto text-gray-800"
      >
        <a
          class="hover:text-point3"
          href="https://github.com/testId"
        >
          <svg
            class="w-8 h-8"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 496 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
            />
          </svg>
        </a>
        <button
          aria-label="email"
          class="hover:text-point3"
        >
          <svg
            class="w-8 h-8"
            fill="currentColor"
            height="1em"
            role="img"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 5.457v13.909c0 .904-.732 1.636-1.636 1.636h-3.819V11.73L12 16.64l-6.545-4.91v9.273H1.636A1.636 1.636 0 0 1 0 19.366V5.457c0-2.023 2.309-3.178 3.927-1.964L5.455 4.64 12 9.548l6.545-4.91 1.528-1.145C21.69 2.28 24 3.434 24 5.457z"
            />
          </svg>
        </button>
        <button
          aria-label="phone"
          class="hover:text-point3"
        >
          <svg
            class="w-8 h-8"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M391 480c-19.52 0-46.94-7.06-88-30-49.93-28-88.55-53.85-138.21-103.38C116.91 298.77 93.61 267.79 61 208.45c-36.84-67-30.56-102.12-23.54-117.13C45.82 73.38 58.16 62.65 74.11 52a176.3 176.3 0 0 1 28.64-15.2c1-.43 1.93-.84 2.76-1.21 4.95-2.23 12.45-5.6 21.95-2 6.34 2.38 12 7.25 20.86 16 18.17 17.92 43 57.83 52.16 77.43 6.15 13.21 10.22 21.93 10.23 31.71 0 11.45-5.76 20.28-12.75 29.81-1.31 1.79-2.61 3.5-3.87 5.16-7.61 10-9.28 12.89-8.18 18.05 2.23 10.37 18.86 41.24 46.19 68.51s57.31 42.85 67.72 45.07c5.38 1.15 8.33-.59 18.65-8.47 1.48-1.13 3-2.3 4.59-3.47 10.66-7.93 19.08-13.54 30.26-13.54h.06c9.73 0 18.06 4.22 31.86 11.18 18 9.08 59.11 33.59 77.14 51.78 8.77 8.84 13.66 14.48 16.05 20.81 3.6 9.53.21 17-2 22-.37.83-.78 1.74-1.21 2.75a176.49 176.49 0 0 1-15.29 28.58c-10.63 15.9-21.4 28.21-39.38 36.58A67.42 67.42 0 0 1 391 480z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="custom-shape-divider bottom-to-top undefined"
    >
      <svg
        data-name="Layer 1"
        preserveAspectRatio="none"
        viewBox="0 0 1200 120"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          class="shape-fill"
          d="M0,0V46.29c47.79,22.2,103.59,32.17,158,28,70.36-5.37,136.33-33.31,206.8-37.5C438.64,32.43,512.34,53.67,583,72.05c69.27,18,138.3,24.88,209.4,13.08,36.15-6,69.85-17.84,104.45-29.34C989.49,25,1113-14.29,1200,52.47V0Z"
          opacity=".25"
        />
        <path
          class="shape-fill"
          d="M0,0V15.81C13,36.92,27.64,56.86,47.69,72.05,99.41,111.27,165,111,224.58,91.58c31.15-10.15,60.09-26.07,89.67-39.8,40.92-19,84.73-46,130.83-49.67,36.26-2.85,70.9,9.42,98.6,31.56,31.77,25.39,62.32,62,103.63,73,40.44,10.79,81.35-6.69,119.13-24.28s75.16-39,116.92-43.05c59.73-5.85,113.28,22.88,168.9,38.84,30.2,8.66,59,6.17,87.09-7.5,22.43-10.89,48-26.93,60.65-49.24V0Z"
          opacity=".5"
        />
        <path
          class="shape-fill"
          d="M0,0V5.63C149.93,59,314.09,71.32,475.83,42.57c43-7.64,84.23-20.12,127.61-26.46,59-8.63,112.48,12.24,165.56,35.4C827.93,77.22,886,95.24,951.2,90c86.53-7,172.46-45.71,248.8-84.81V0Z"
        />
      </svg>
    </div>
  </section>
</DocumentFragment>
`;
